src/Microsoft.TestPlatform.AdapterUtilities/ManagedNameUtilities/ManagedNameHelper.Reflection.cs

```diff
@@ -266,14 +268,169 @@ private static void AppendTypeString(StringBuilder b, Type type, bool closedType
+            }
+        }
+
+        private static void NormalizeString(StringBuilder b, string name)
+        {
+            b.Append("'");
+            for (int i = 0; i < name.Length; i++)
+            {
+                char c = name[i];
+                if (NeedsEscaping(c, i))
+                {
+                    if (c == '\\' || c == '\'')
+                    {
+                        // var encoded = Convert.ToString(((uint)c), 16);
+                        // b.Append("\\u");
+                        // b.Append('0', 4 - encoded.Length);
+                        // b.Append(encoded);
+
+                        b.Append("\\");
+                        b.Append(c);
+                        continue;
+                    }
+                }
+
+                b.Append(c);
+            }
+            b.Append("'");
+        }
+
+        private static bool IsNormalized(string s)
+        {
+            for (int i = 0; i < s.Length; i++)
+            {
+                if (NeedsEscaping(s[i], i) && s[i] != '.')
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static bool NeedsEscaping(char c, int pos)
+        {
+            if (pos == 0 && char.IsDigit(c))
+            {
+                return true;
+            }
+
+            if (c == '_'
+              || char.IsLetterOrDigit(c) // Lu, Ll, Lt, Lm, Lo, or Nl 
+              )
+            {
+                return false;
+            }
+
+            var category = CharUnicodeInfo.GetUnicodeCategory(c);
+            if (category == UnicodeCategory.NonSpacingMark        // Mn
+             || category == UnicodeCategory.SpacingCombiningMark  // Mc
+             || category == UnicodeCategory.ConnectorPunctuation  // Pc
+             || category == UnicodeCategory.Format)               // Cf
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static int AppendNestedTypeName(StringBuilder b, Type type)
+        {
+            var outerArity = 0;
             if (type.IsNested)
             {
-                AppendNestedTypeName(b, type.DeclaringType);
+                outerArity = AppendNestedTypeName(b, type.DeclaringType);
                 b.Append('+');
             }
-            b.Append(type.Name);
+
+            var typeName = type.Name;
+            var stars = 0;
+            if (type.IsPointer)
+            {
+                for (int i = typeName.Length - 1; i > 0; i--)
+                {
+                    if (typeName[i] != '*')
+                    {
+                        stars = typeName.Length - i - 1;
+                        typeName = typeName.Substring(0, i + 1);
+                        break;
+                    }
+                }
+            }
+
+            var info = type.GetTypeInfo();
+            var arity = info.IsGenericType == false ? 0
```

Comment:
Consider splitting this conditional expression in `if-else if` chains or add proper indentation like this:

```
var arity = !info.IsGenericType
    ? 0
    : info.GenericTypeParameters.Length > 0
        ? info.GenericTypeParameters.Length
        : info.GenericTypeArguments.Length;
```